tests/test_markers.py

```diff
@@ -303,6 +303,46 @@ def test_extra_with_no_extra_in_environment(self):
                 {"extra": "security"},
                 True,
             ),
+            (
+                "extra == 'security'",
+                {"extra": None},
+                False,
+            ),
+            (
+                "extra == '3.7.0'",
+                {"extra": None},
+                False,
+            ),
+            (
+                "extra == '3.7.0'",
+                {"extra": "foobar"},
+                False,
+            ),
+            (
+                "extra == '3.7.0'",
+                {"extra": "3.7.0"},
+                True,
+            ),
+            (
+                "extra == '3.7.0'",
+                {"extra": "3.7"},
+                True,
+            ),
+            (
+                "extra == '3.7.0'",
+                {"extra": "3.7.0.0"},
+                True,
+            ),
+            (
+                "extra == '3.7.0'",
+                {"extra": "3.7.0+youaregreat"},
+                True,
```

Comment:
After looking more into the code, I realized that having "Variable Types" wouldn't be as hard as I thought. I have a general idea of how I could implement that.

Here's my suggestion:
- Merge this PR, as it fixes a bug that is live, which is that if a packages provides extras that are valid version specifier, the package is uninstalable (pointing back at  pypa/pip#4356). 
This PR does not actually bring the change that introduce the questionable extra behaviour: If someone, right now with the previous code, creates a `Marker` object like `extra == 3.7.0+youaregreat` and evaluates it with an environment like `{'extra': '3.7.0'}`, it would have the behaviour I simply explicited in the tests. This simply never really occured "in the wild" before as the principal (if not only?) use of that is through setuptools/pip, which crashed before that by passing a `None` and crashed, which is now fixed.

- Open an issue and re-launch the discussion on "Is that really the behaviour we intended with PEP 508?", as this PR pointed out there was an edge case with the extras possible values. We can then make the required changes/clarifications to PEP 508 and make the appropriate change to the code to respect the newly defined and written behaviour, if needed. I'll gladly join that discussion, and I'll most likely be ready to help by contributing with a PR for the changes.

If we could come to a conclusion really fast on the previously intended PEP 508 behaviour, I would gladly put it in this PR right away, but as we probably all have lots of things to do, I feel like that discussion will take quite some time, and it would be unfortunate to block a fix to a live bug for that.

Thanks again for your time!
@dstufft @xavfernandez